expr/functions/integralByInterval/function.go

```diff
@@ -44,28 +44,25 @@ func (f *integralByInterval) Do(ctx context.Context, e parser.Expr, from, until
 		return nil, err
 	}
 	bucketSize := int64(bucketSizeInt32)
-	bucketSizeStr := e.Arg(1).StringValue()
+	intervalString, err := e.GetStringArg(1)
+	if err != nil {
+		return nil, err
+	}
 
 	startTime := from
 	results := make([]*types.MetricData, len(args))
 	for j, arg := range args {
 		current := 0.0
 		currentTime := arg.StartTime
 
-		name := "integralByInterval(" + arg.Name + ",'" + bucketSizeStr + "')"
-		result := &types.MetricData{
-			FetchResponse: pb.FetchResponse{
-				Name:              name,
-				Values:            make([]float64, len(arg.Values)),
-				StepTime:          arg.StepTime,
-				StartTime:         arg.StartTime,
-				StopTime:          arg.StopTime,
-				XFilesFactor:      arg.XFilesFactor,
-				PathExpression:    name,
-				ConsolidationFunc: arg.ConsolidationFunc,
-			},
-			Tags: arg.Tags,
-		}
+		name := fmt.Sprintf("integralByInterval(%s,'%s')", arg.Name, e.Args()[1].StringValue())
+		result := arg.CopyLink()
+		result.Name = name
+		result.PathExpression = name
+		result.Values = make([]float64, len(arg.Values))
+
+		result.Tags["integralByInterval"] = intervalString
```

Comment:
Tags need be copied before append new tags or modify existing

Reply:
I think they are being copied, look at the `CopyLink` call a few lines above:


https://github.com/grafana/carbonapi/blob/d511d4c9150c851de8e77cd4f3014071ad754e61/expr/functions/integralByInterval/function.go#L59-L64

Reply:
Yes, you are right.